pkg/binary/envoy/run.go

```diff
@@ -46,32 +45,50 @@ func (r *Runtime) RunPath(path string, args []string) error {
 		return fmt.Errorf("unable to stat %q: %v", path, err)
 	}
 
-	ctx, cancel := context.WithCancel(context.Background())
-	r.ctx = ctx
-
-	// #nosec -> users can run whatever binary they like!
-	r.cmd = exec.Command(path, args...)
-	r.cmd.Dir = r.WorkingDir
-	r.cmd.Stdout = r.IO.Out
-	r.cmd.Stderr = r.IO.Err
-	r.cmd.SysProcAttr = sysProcAttr()
+	// We can't use CommandContext even if that seems correct here. The reason is that we need to invoke preTerminate
+	// handlers, and they expect the process to still be running. For example, this allows admin API hooks.
+	cmd := exec.Command(path, args...) // #nosec -> users can run whatever binary they like!
+	cmd.Dir = r.WorkingDir
+	cmd.Stdout = r.IO.Out
+	cmd.Stderr = r.IO.Err
+	cmd.SysProcAttr = sysProcAttr()
+	r.cmd = cmd
 
 	r.handlePreStart()
 
-	go r.runEnvoy(cancel)
+	if cmd.Stdout == nil {
+		cmd.Stdout = os.Stdout
+	}
+	if cmd.Stderr == nil {
+		cmd.Stderr = os.Stderr
+	}
 
-	r.waitForTerminationSignals()
-	r.handleTermination()
-	cancel() // this should never actually do anything but lets avoid any future context leaks
+	log.Infof("Envoy command: %v", cmd.Args)
+	if err := cmd.Start(); err != nil {
+		return fmt.Errorf("unable to start Envoy process: %w", err)
+	}
+
+	waitCtx, waitCancel := context.WithCancel(context.Background())
+	sigCtx, sigCancel := signal.NotifyContext(waitCtx, syscall.SIGINT, syscall.SIGTERM)
+	defer waitCancel()
+	r.fakeInterrupt = sigCancel
+
+	go waitForExit(cmd, waitCancel) // waits in a goroutine. We may need to kill the process if a signal occurs first.
 
-	// Block until the Envoy process and termination handler are finished cleaning up
-	r.wg.Wait()
+	// Block until we receive SIGINT or are canceled because Envoy has died
+	<-sigCtx.Done()
 
-	// Tar up the debug data and clean up
-	if err := archiver.Archive([]string{r.DebugStore()}, r.DebugStore()+".tar.gz"); err != nil {
-		return fmt.Errorf("unable to archive debug store directory %v: %v", r.DebugStore(), err)
+	if cmd.ProcessState != nil {
+		log.Infof("Envoy process (PID=%d) terminated prematurely", cmd.Process.Pid)
+		return r.handlePostTermination()
 	}
-	return os.RemoveAll(r.DebugStore())
+
+	r.handleTermination()
+
+	// Block until the process is complete. This ensures file descriptors are closed.
+	<-waitCtx.Done()
```

Comment:
this is what now enforces clean stop prior to invoking any post termination hooks (currently only log and archive hooks)